### DISCUSSION TITLE
How to change state variable of functionComponent in cypress?

### DISCUSSION THREAD

**Discussion by vikasm98** [cypress-io/cypress#25656] [Questions and Help]
Hello
I have a state variable declared from useState() hook, as `const [showComponent, setShowComponent] = useState(false);`

Here I would like to change  showComponent to true using setShowComponent(true) in cypress spec file before running assertions in cypress component testing.
Please help!

**Comment by portyu9** [comment]
Hi @vikasm98 

You cannot directly call a function component's internal `setShowComponent()` from a Cypress spec.

With:

```js
const [showComponent, setShowComponent] = useState(false)
```

both the state and its setter belong to that particular React component render. Unlike a class component, there is no component instance that Cypress can reach into and modify.

The preferred test is to perform whatever user action normally changes that state:

```js
cy.mount(<MyComponent />)

cy.get('[data-cy=show-component]').click()

cy.get('[data-cy=component]').should('be.visible')
```

If there is deliberately no UI action that can create the state you need, make the state controllable from outside the component — for example through a prop, context, or store — and initialize that dependency when mounting the component.

For example, if the component supports an initial value:

```js
cy.mount(<MyComponent initialShowComponent={true} />)
```

The important point is that Cypress should control the component through its **public interface**, rather than trying to invoke a `useState()` setter from outside the function component.

Please mark this answer as accepted if it helped, thank you.
